pbs_light.go

```diff
@@ -794,6 +800,34 @@ func serve(cfg *config.Configuration) error {
 		return fmt.Errorf("Prebid Server could not load data cache: %v", err)
 	}
 
+	// Validate currency set in config
+	if cfg.AdServerCurrency != "" {
+		currency.MustParseISO(cfg.AdServerCurrency)
+	}
+
+	// Load latest currency conversion rates in blocking call
+	currencyLoadStopSignal := make(chan os.Signal, 1)
+	signal.Notify(currencyLoadStopSignal, syscall.SIGTERM, syscall.SIGINT)
+	var currencyRates interface{}
```

Comment:
so... this code is extremely tricky to write. There are a few bugs here:

1. Assignments [are not atomic](https://stackoverflow.com/questions/21447463/is-assigning-a-pointer-atomic-in-golang) in Go... so this is a race condition. That's actually why `atomic.Value`exists--`Store()` _replaces_ assignments in a threadsafe way.

2. Below, you're passing `currencyRates.([]byte)` into `NewExchange`. The fact that `CurrencyRates.Store()` gets called _after_ that won't affect the value which you passed to the Exchange.

And also a few structural issues. These are more subjective, but directly tied to the bugs:

1. We should never write code with package-level mutable variables in it. It's extremely difficult to write proper tests for any code which uses them, which makes the code very buggy. As an example, consider how you would write a test to make sure that the `Exchange` uses updated currency values. It's a massive headache, if not impossible.

2. I don't think we've communicated this well to Rubicon... but our goal is to _remove_ code from this file, not add to it. In an ideal world, this file will become a bunch of `New` calls and a `server.Listen()`.

--------------------------------------------------------------------------------------------

Specifically, I would recommend moving this feature into a `currency` module. API would be something like this:

```
func NewCurrencyConverter() CurrencyConverter {
  // Block waiting for the initial HTTP call
  // Make a CurrencyConverter implementation with the atomic.Value as a field
  // Create a goroutine which updates the atomic.Value periodically
  // return the object
}

type CurrencyConverter interface {
  Convert(price float64, old string, new string) float64 
}
```

Given those, it should be possible to build a CurrencyConverter with fairly simple unit tests that update the values mid-test, and then make sure it returns those updated values.

Reply:
good catch on issues 1 and 2. I've updated the code. I'm a newbie with atomic values in go so please let me know if this works.
you make good points on the structure of the code, but I'm going to hold off on making any of those changes for  now since I'm a bit overloaded on other tasks at the moment. but if there's anything that is not working please do point it out.

Reply:
That's fine, I understand.

Honestly... I won't feel comfortable merging it without some unit tests. I could see _these_ bugs, but I'm not an expert either, and race conditions here would cause the entire process to crash under higher traffic. With the interface I suggested, we could make a test which updates the data & calls `Convert` in parallel, and run it under the [race detector](https://blog.golang.org/race-detector) to make sure it won't cause problems.

Since the `serve` function is inherently untestable, it'll need _some_ sort of restructuring to make it testable first. If you're overloaded, I can check in with @mkendall07 and see if he's willing to devote some of my time to it. Otherwise I don't think anyone on our end would mind waiting until the other stuff settles down and there's time to devote to it.